web/src/main/java/com/bazaarvoice/emodb/web/resources/databus/DatabusResource1.java

```diff
@@ -394,4 +417,93 @@ private SubjectDatabus getClient(BooleanParam partitioned) {
+        Collection<String> splits = _dataStore.getSplits(table, batchSize.get());
+        Predicate<Map<String, Object>> predicate = getDatePredicate(date);
+        splits.parallelStream().forEach(split -> {
+            Iterable<Map<String, Object>> documents = DataStoreStreaming.getSplit(_dataStore, table, split, true, consistency.get());
+            List<UpdateRef> updateRefs = StreamSupport.stream(documents.spliterator(), true)
+                    .filter(predicate)
+                    .map(document -> getUpdateRef(document))
+                    .collect(Collectors.toList());
+
+            _eventWriterRegistry.getDatabusWriter().writeEvents(updateRefs);
+            Iterators.concat(result, updateRefs.iterator());
+        });
+        return streamingIterator(result);
+    }
+
+    private static Predicate<Map<String, Object>> getDatePredicate(Date date) {
+        Predicate<Map<String, Object>> datePredicate = document -> Intrinsic.getLastUpdateAt(document).compareTo(date) >= 0;
+        Predicate<Map<String, Object>> alwaysTruePredicate = document -> true;
+        return date == null ? alwaysTruePredicate : datePredicate;
+    }
+
+    @POST
+    @Path ("_write-key-event/{table}/{key}")
+//    @RequiresPermissions ("sor|read|{table}, system|raw_databus")
+    @Timed (name = "bv.emodb.sor.DatabusResource1.writeEvent", absolute = true)
+    @ApiOperation (value = "Replay operation.",
+            notes = "Replay operation.",
+            response = UpdateRef.class
+    )
+    public UpdateRef sendEvent(@PathParam ("table") String table,
+                               @PathParam ("key") String key,
+                               @QueryParam ("consistency") @DefaultValue ("STRONG") ReadConsistencyParam consistency,
+                               @QueryParam ("since") InstantParam sinceParam,
+                               @Authenticated Subject subject) {
+        checkArgument(!Strings.isNullOrEmpty(table), "table is required");
+        checkArgument(!Strings.isNullOrEmpty(key), "key is required");
+
+        Instant since = (sinceParam == null) ? null : sinceParam.get();
+        Date date = since != null ? Date.from(since) : null;
+        Map<String, Object> document = _dataStore.get(table, key, consistency.get());
+        UpdateRef updateRef = null;
+        // send events only if sinceParam isn't specified, or
+        // document had been created before sinceParam
+        if (getDatePredicate(date).test(document)) {
+            updateRef = getUpdateRef(document);
+            _eventWriterRegistry.getDatabusWriter().writeEvents(Collections.singleton(updateRef));
+        }
+        return updateRef;
+    }
+
+    private static UpdateRef getUpdateRef(final Map<String, Object> document) {
+        Set<String> tags = (Set<String>) document.get(UpdateRef.TAGS_NAME);
+        //determine changedId base on lastUpdateAt timestamp
+        UUID changeId = TimeUUIDs.uuidForTimestamp(Intrinsic.getLastUpdateAt(document));
```

Comment:
IMO the changeId for the new event need not be the last update time of the doc. Theoretically an event with the same timestamp was already generated with the last delta for that document. Its’ not about that, but looks like u are generating a UUID which may be differing. so may be that is not working.  I prefer creating a new time, TimeUUIDs.newUUID() which is beyond the FCT which will make sure the databus event is not skipped. so may be use random 7 days prior to now() which most of the cases will be behind FCT or Get the current FCT value and subtract 5 minutes from that or something better. We can also have a parameter coming in for this timestamp or the time we are subtracting in this case; remember the endpoint is only used by admins. so we can be flexible in tweaking this …..

Reply:
also date should be within `~firstUpdateAt` and `FCT - something` for each document, right?